### PR TITLE
Add support for dense rank.

### PR DESCRIPTION
**Pull request by BillWhite**
In this commit we add support for dense rank.  We also refactor the
parsing of windowed aggregates to be less rank-centric and more general.

Also, with this commit, we repair some errors in the build system.
These caused unnecessary and misleading error messages, and are more
robust to GCC versions.

### CODE REVIEW

File: build.py [hunk 1]
```diff
@@ -54,13 +54,24 @@
...
     if (CTX.compilerMajorVersion() >= 4):
         CTX.CPPFLAGS += " -Wno-deprecated-declarations  -Wno-unknown-pragmas"
-	if (CTX.compilerMinorVersion() == 6):
-	    CTX.CPPFLAGS += " -Wno-unused-but-set-variable"
-	if (CTX.compilerMinorVersion() == 9):
-            CTX.CPPFLAGS += " -Wno-float-conversion -Wno-unused-but-set-variable -Wno-unused-local-typedefs"
-        elif (CTX.compilerMinorVersion() == 8):
-	    CTX.CPPFLAGS += " -Wno-conversion -Wno-unused-but-set-variable -Wno-unused-local-typedefs"
-
+    # GCC 4 warning disablement options
```
Comment (BillWhite) [inline build.py]: The old code confused GCC4 minor versions with GCC5 minor versions.  It only worked by accident, because the two sets of minor version numbers are disjoint.  That won't happen forever.  Besides, organizing this by individual warning disablements and not by compiler version is simpler, I think.

File: buildtools.py [hunk 2]
```diff
@@ -315,12 +315,12 @@ def buildThirdPartyTools(CTX, makefile):
     makefile.write('.PHONY: build-pcre2 configure-pcre2 unpack-pcre2\n')
     makefile.write('build-pcre2: configure-pcre2\n')
     makefile.write('\tif [ ! -f "${PCRE2_INSTALL}/lib/libpcre2-8.a" ] ; then \\\n')
-    makefile.write('\t  @echo Building PCRE2 ; \\\n')
+    makefile.write('\t  echo Building PCRE2 ; \\\n')
```
Comment (BillWhite) [inline buildtools.py]: Using @echo in the makefile caused a spurious error message at compile time.  This is harmless, but sloppy looking.

File: src/frontend/org/voltdb/planner/AbstractParsedStmt.java [hunk 3]
```diff
@@ -326,47 +326,61 @@ public AbstractExpression parseExpressionTree(VoltXMLElement root) {
         return expr;
     }
 
+    private interface XMLElementExpressionParser {
+        AbstractExpression parse(AbstractParsedStmt stmt,
+                                 VoltXMLElement element);
+    }
+
+    private static Map<String, XMLElementExpressionParser> m_exprParsers;
```
Comment (BillWhite) [inline src/frontend/org/voltdb/planner/AbstractParsedStmt.java]: These used to be a big if statement in the function parseExpressionNode.  This lookup table has to be faster, and this seems like it's in the inner loop of all VoltXMLElement parsing.

File: src/frontend/org/voltdb/planner/ParsedSelectStmt.java [hunk 4]
```diff
@@ -2196,33 +2199,49 @@ public String calculateContentDeterminismMessage() {
     public boolean isDML() { return false; }
 
     /**
-     * Return true iff all the windowed expressions have a table partition column
-     * in their partition by list.  Since we only support one windowed expression
-     * now, this returns true if there are no windowed expressions.
+     * Return true iff all the windowed partition expressions have a table partition column
+     * in their partition by list, and if there is one such windowed partition expression.  If there are
+     * no windowed expressions, we return false.  Note that there can only be one windowed
+     * expression currently, so this is more general than it needs to be.
```
Comment (BillWhite) [inline src/frontend/org/voltdb/planner/ParsedSelectStmt.java]: The comment on this function was wrong, and that led me to believe that the function could fail in some edge cases.  So I tried to make it more robust.

File: src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionRank.java [hunk 5]
```diff
@@ -1,167 +0,0 @@
-/* Copyright (c) 2001-2009, The HSQL Development Group
```
Comment (BillWhite) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionRank.java]: This class, and the file containing it, were renamed to be ExpressionWindowed.

File: tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java [hunk 6]
```diff
@@ -253,36 +253,42 @@ public void testRank_NON_UNIQUE() throws NoConnectionsException, IOException, Pr
...
-    final int colC      = 3;
-    final int colR_A    = 4;
-    final int colR_AA   = 5;
+    final int colA          = 0;
+    final int colAA         = 1;
+    final int colB          = 2;
+    final int colC          = 3;
+    final int colR_A        = 4;
+    final int colR_AA       = 5;
+    final int colR_dense    = 6;
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/regressionsuites/TestWindowedAggregateSuite.java]: Should there be a colR_dense_AA and associated test queries?

File: src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+ *
+ * @author Xin Jia
+ */
+public class ExpressionWindowed extends Expression {
+    private List<Expression> m_partitionByList;
+    private SortAndSlice     m_sortAndSlice;
+    private boolean          m_isDistinctAggregate;
+
+    ExpressionWindowed(int tokenT,
+                       Expression aggExprs[],
```
Comment (paulmartel) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java]: I'm a little put off by the addition of the windowed agg args code as dead code at this premature stage.
The logic "seems correct" but it is not tested and not easily testable until we support actual windowed aggs.
It would make more sense to review and merge this code at that stage.
Though I guess I should have objected to the scope creep at standup.

File: src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+                       boolean isDistinct,
```
Comment (paulmartel) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java]: There's a subtle disconnect between isDistinct and m_isDistinctAggregate here -- the kind of thing that happens and can be easily missed in untestable dead (for now?) code.

File: src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+                orderby.attributes.put("decending", isDecending ? "true": "false");
+                rxe.children.add(orderby);
+            }
+        }
+
+        if (nodes.length > 0) {
+            for (Expression expr : nodes) {
+                winargs.children.add(expr.voltGetXML(context, null));
+            }
+        }
```
Comment (paulmartel) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java]: Make this more consistent with the  other methods by processing nodes before m_partitionByList

File: src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+            sb.append(e.getSQL());
+        }
+        sb.append(") ")
+          .append(Tokens.T_OVER + " (");
+        if (m_partitionByList.size() > 0) {
+            sb.append(Tokens.T_PARTITION + ' ' + Tokens.T_BY + ' ');
+            for (int idx = 0; idx < m_partitionByList.size(); idx += 1) {
+                Expression expr = m_partitionByList.get(idx);
+                sb.append(expr.getSQL());
+                if (idx < m_partitionByList.size()-1) {
```
Comment (paulmartel) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java]: I try to handle separator logic like this with a more terse reassigned prefix idiom:

        // No prefix for the first element.
        String prefix = "";
        for(...) {
            ...
            sb.append(prefix).append(...);
            // Prefix any subsequent element.
            prefix = ", ";
        }

BTW, we don't usually put a space before a closing paren, anyway.

File: build.py [hunk 11]
```diff
@@ -84,19 +95,32 @@
...
+    elif (CTX.compilerMajorVersion() == 5):
+        CTX.CXX_VERSION_FLAG = "c++11"
     else:
-	CTX.CXX_VERSION_FLAG ="--std=c++11"
-	print("Building with C++11")
+        print("GCC Version %d.%d.%d is too new\n"
+               % (CTX.compilerMajorVersion(),
+                  CTX.compilerMinorVersion(),
+                  CTX.compilerPatchLevel()))
+        sys.exit(-1)
```
Comment (BillWhite) [inline build.py]: Ok.

File: build.py [hunk 12]
```diff
@@ -84,19 +95,32 @@
...
 elif CTX.compilerName() == 'clang':
-    CTX.CXX_VERSION_FLAG="--std=c++11"
-CTX.CPPFLAGS += " " + CTX.CXX_VERSION_FLAG
+    CTX.CXX_VERSION_FLAG="c++11"
+else:
+    print("Unknown compiler %s" % CTX.compilerName())
+    sys.exit(-1)
```
Comment (BillWhite) [inline build.py]: Ok.

File: src/ee/executors/aggregateexecutor.cpp [hunk 13]
```diff
@@ -613,15 +613,16 @@ inline Agg* getAggInstance(Pool& memoryPool, ExpressionType agg_type, bool isDis
...
-    {
-        char message[128];
-        snprintf(message, sizeof(message), "Unknown aggregate type %d", agg_type);
-        throw SerializableEEException(VOLT_EE_EXCEPTION_TYPE_EEEXCEPTION, message);
-    }
+        {
+            char message[128];
+            snprintf(message, sizeof(message), "Unknown aggregate type %d", agg_type);
+            throw SerializableEEException(VOLT_EE_EXCEPTION_TYPE_EEEXCEPTION, message);
+        }
```
Comment (BillWhite) [inline src/ee/executors/aggregateexecutor.cpp]: They were indented in the other cases, but not in this one.  I thought uniformity was better.

File: src/ee/executors/aggregateexecutor.cpp [hunk 14]
```diff
@@ -837,7 +840,7 @@ TableTuple AggregateExecutorBase::p_execute_init(const NValueArray& params,
     m_inProgressGroupByKeyTuple.move(NULL);
 
     char * storage = reinterpret_cast<char*>(
-            m_memoryPool.allocateZeroes(schema->tupleLength() + TUPLE_HEADER_SIZE));
+                                             m_memoryPool.allocateZeroes(schema->tupleLength() + TUPLE_HEADER_SIZE));
```
Comment (BillWhite) [inline src/ee/executors/aggregateexecutor.cpp]: Ok.  Emacs broke it, not me.

File: src/hsqldb19b3/org/hsqldb_voltpatches/Expression.java [hunk 15]
```diff
@@ -1527,7 +1527,7 @@ public void collectAllColumnExpressions(HsqlList set) {
         prototypes.put(OpTypes.MATCH_UNIQUE_SIMPLE,  null); // not yet supported ExpressionLogical
         prototypes.put(OpTypes.MATCH_UNIQUE_PARTIAL, null); // not yet supported ExpressionLogical
         prototypes.put(OpTypes.MATCH_UNIQUE_FULL,    null); // not yet supported ExpressionLogical
-        // aggregate functions
+        // grouped aggregate functions
```
Comment (BillWhite) [inline src/hsqldb19b3/org/hsqldb_voltpatches/Expression.java]: Well, I changed it to unwindowed, but I still think it's grouped.  If there are no GROUP BY expressions the result is one big group for the purposes of the aggregate calculation.

File: src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java (hunk 7 above)
Comment (BillWhite) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java]: It's not that premature.  We will support max and min before too very long.  Adding more windowed aggregates seems to be in the near term future.

File: src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java (hunk 8 above)
Comment (BillWhite) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java]: Ok, I'll watch out for that.

File: src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java (hunk 10 above)
Comment (BillWhite) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ExpressionWindowed.java]: ok.

File: src/hsqldb19b3/org/hsqldb_voltpatches/ParserDQL.java [hunk 16]
```diff
@@ -1503,6 +1507,17 @@ private Expression readAggregateExpression(int tokenT) {
...
+        // parameters.  So, just return null here.  The caller
+        // will know what to do with this.
+        switch (tokenT) {
+        case Tokens.RANK:
+        case Tokens.DENSE_RANK:
+        // No current support for WINDOWED_PERCENT_RANK or WINDOWED_CUME_DIST.
+        // case Tokens.PERCENT_RANK:
+        // case Tokens.CUME_DIST:
+            return null;
+        }
```
Comment (BillWhite) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ParserDQL.java]: Yes, RANK() DISTINCT is not standard.  I need to add negative cases for this.

File: src/hsqldb19b3/org/hsqldb_voltpatches/ParserDQL.java [hunk 17]
```diff
@@ -1580,19 +1600,36 @@ private Expression readRank(boolean isPercent) {
...
+        // aggregate expression as a non-windowed expression.  We do
+        // care about its parameters and whether it's specified as
+        // unique though.
+        assert(aggExpr == null || aggExpr instanceof ExpressionAggregate);
+        Expression nodes[];
+        boolean isDistinct;
+        if (aggExpr != null) {
+            ExpressionAggregate winAggExpr = (ExpressionAggregate)aggExpr;
+            nodes = winAggExpr.nodes;
+            isDistinct = winAggExpr.isDistinctAggregate;
```
Comment (BillWhite) [inline src/hsqldb19b3/org/hsqldb_voltpatches/ParserDQL.java]: For RANK it's dead code, but for some other aggregate it would not be.

**Comment by BillWhite** [comment]
Squashed elsewhere.
